### ISSUE TITLE
[Question] Understanding How Values in PPO.predict() are Produced

### ISSUE THREAD

**Issue by tancred0** [DLR-RM/stable-baselines3#2126]
### ❓ Question

I'm trying to understand the internal mechanics of PPO.predict() with a custom environment that uses VecNormalize. Here's my specific issue:

## Current Understanding

- My observations are scaled through VecNormalize using running mean and standard deviation
- I believe I need to use PPO.policy.extract_features to convert my MultiInputPolicy into a flattened array
- Then, using the mlp_extractor, I should get the input to my policy network
- This process should give me the (potentially unscaled?) outputs

My Current Code
I've written a function to trace through the prediction process, but the output significantly differs from what model.predict() produces:

```python
obs_normalizer = vec_env.obs_rms
def trace_prediction(model, observation):
    # Let the model preprocess the observation properly
    processed_obs = model.policy.obs_to_tensor(observation)[0]
    
    scaled_obs = {}
    with torch.no_grad():
        for key in observation.keys():
            if key in obs_normalizer:  # Wrong: obs_mean is not a dict
                obs_value = np.array(observation[key])
                obs_normalized = (obs_value - obs_normalizer[key].mean) / np.sqrt(obs_normalizer[key].var)
                scaled_obs[key] = torch.FloatTensor(obs_normalized.reshape(1, -1)).to(model.device)
            else:
                print("key not in obs_mean:", key)
        print(scaled_obs)
        # Extract features
        features = model.policy.extract_features(scaled_obs)
        # Get latent policy
        latent_pi, latent_vf = model.policy.mlp_extractor(features)
        # Get action distribution
        action_dist = model.policy.action_dist
        action_raw = model.policy.action_net(latent_pi)
        
        # Handle distribution if needed
        if hasattr(model.policy, 'action_dist'):
            # For stochastic policies
            action_dist = model.policy.action_dist.proba_distribution(action_raw, log_std=model.policy.log_std)
            actions = action_dist.get_actions(deterministic=True)
        else:
            # For deterministic policies
            actions = action_raw
        # convert to numpy
        actions_np = actions.detach().cpu().numpy()
        
        # un-scale if that hook exists
        if hasattr(model.policy, 'unscale_action'):
            actions_scaled = model.policy.unscale_action(actions_np)
        return actions_np.flatten(), actions_scaled.flatten()

```
## My Question
What's wrong with this code? How can I deterministically get results that match model.predict() by passing my observations through the built-in neural network?
My repository is too large to share in its entirety, but I'm looking for guidance on the conceptual or implementation errors in my approach.

Example difference between PPO.predict and the unscaled outputs.

<img width="317" alt="Image" src="https://github.com/user-attachments/assets/8a1d0019-55ed-48c3-927b-889ea29cf37e" />

### Checklist

- [x] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [x] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [x] If code there is, it is [minimal and working](https://github.com/DLR-RM/stable-baselines3/issues/982#issuecomment-1197044014)
- [x] If code there is, it is formatted using the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by araffin** [comment]
Hello, 
to understand the motivation, why do you need to recreate the predict method?

**Comment by tancred0** [comment]
Hi,

I am doing this as part of my master thesis.

When training is finished, my agent performs pretty well on average but sometimes he performs counterintuitive actions. Thus I would like to know if the problem exists inside my codebase (data, rollouts, maybe weird normalization) and I thought that looking at the inner workings would help to run sub-analyses.

**Comment by araffin** [comment]
> my agent performs pretty well on average but sometimes he performs counterintuitive actions.

That might be due to the stochastic nature of your env (or are you testing with `deterministic=False` ?)
I would recommend using the [RL Zoo](https://github.com/DLR-RM/rl-baselines3-zoo) to avoid any mismatch between training and testing.


As you probably found, the main logic of predict is here:
https://github.com/DLR-RM/stable-baselines3/blob/c1e503c21e068668da5072d91d0408da0d7c0a17/stable_baselines3/common/policies.py#L365-L379

In the case of PPO, clipping is used (because it doesn't use a [squashed Gaussian](https://araffin.github.io/post/sac-massive-sim/)).
The `_predict()` is 
https://github.com/DLR-RM/stable-baselines3/blob/c1e503c21e068668da5072d91d0408da0d7c0a17/stable_baselines3/common/policies.py#L717
and
https://github.com/DLR-RM/stable-baselines3/blob/c1e503c21e068668da5072d91d0408da0d7c0a17/stable_baselines3/common/policies.py#L750-L752

where the feature extraction is defined as https://github.com/DLR-RM/stable-baselines3/blob/c1e503c21e068668da5072d91d0408da0d7c0a17/stable_baselines3/common/policies.py#L130-L131

**Comment by tancred0** [comment]
thank you very much 🙏🏼 i will check whether i get the same / similar values with the functions
